File: packages/domain/src/consumableLifecycle.ts
export interface ConsumableLifecycleIssue {
  readonly code: "chronology";
  readonly path: string;
  readonly relatedPath: string;
  readonly message: string;
}

type LifecycleValue = Readonly<Record<string, unknown>>;
type ChronologyEdge = readonly [earlier: string, later: string];

const FILM_ROLL_COLLECTION = "app.graycard.instance.filmRoll";
const CHEMISTRY_COLLECTION = "app.graycard.instance.chemistry";

// These edges encode only lifecycle orderings that do not depend on a
// particular lab or darkroom workflow. In particular, a lab scan may arrive
// before the physical negatives, so scannedAt and receivedFromLabAt are not
// comparable.
const FILM_ROLL_EDGES: readonly ChronologyEdge[] = [
  ["loadedAt", "partialAt"],
  ["partialAt", "exposedAt"],
  ["exposedAt", "unloadedAt"],
  ["unloadedAt", "sentToLabAt"],
  ["unloadedAt", "developmentStartedAt"],
  ["sentToLabAt", "developmentStartedAt"],
  ["developmentStartedAt", "developedAt"],
  ["developedAt", "receivedFromLabAt"],
  ["developedAt", "scannedAt"],
  ["receivedFromLabAt", "archivedAt"],
  ["scannedAt", "archivedAt"],
];

// acquiredAt and expiresAt are intentionally excluded: a user may acquire an
// already-open or already-mixed solution, and a use-by date is not an event.
// Replenishment and exhaustion are also unordered because replenishment may
// revive an exhausted solution or precede its eventual exhaustion.
const CHEMICAL_EDGES: readonly ChronologyEdge[] = [
  ["openedAt", "mixedAt"],
  ["mixedAt", "replenishedAt"],
  ["mixedAt", "exhaustedAt"],
  ["acquiredAt", "discardedAt"],
  ["openedAt", "discardedAt"],
  ["mixedAt", "discardedAt"],
  ["replenishedAt", "discardedAt"],
  ["exhaustedAt", "discardedAt"],
];

function transitivePairs(edges: readonly ChronologyEdge[]): readonly ChronologyEdge[] {
  const successors = new Map<string, Set<string>>();
  for (const [earlier, later] of edges) {
    const values = successors.get(earlier) ?? new Set<string>();
    values.add(later);
    successors.set(earlier, values);
  }

  const pairs: ChronologyEdge[] = [];
  for (const earlier of successors.keys()) {
    const seen = new Set<string>();
    const pending = [...(successors.get(earlier) ?? [])];
    while (pending.length) {
      const later = pending.pop()!;
      if (seen.has(later)) continue;
      seen.add(later);
      pairs.push([earlier, later]);
      pending.push(...(successors.get(later) ?? []));
    }
  }
  return pairs;
}

const FILM_ROLL_PAIRS = transitivePairs(FILM_ROLL_EDGES);
const CHEMICAL_PAIRS = transitivePairs(CHEMICAL_EDGES);

function timestamp(value: unknown): number | null {
  if (typeof value !== "string" || !value) return null;
  const parsed = Date.parse(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function collectionPairs(collection: string): readonly ChronologyEdge[] | null {
  if (collection === FILM_ROLL_COLLECTION) return FILM_ROLL_PAIRS;
  if (collection === CHEMISTRY_COLLECTION) return CHEMICAL_PAIRS;
  return null;
}

/**
 * Report chronologically impossible consumable milestones.
 *
 * Missing and malformed timestamps are left to lexicon validation; this check
 * only compares valid timestamps whose lifecycle ordering is unambiguous.
 */
export function validateConsumableLifecycle(collection: string, value: LifecycleValue): ConsumableLifecycleIssue[] {
  const pairs = collectionPairs(collection);
  if (!pairs) return [];

  const issues: ConsumableLifecycleIssue[] = [];
  for (const [earlierField, laterField] of pairs) {
    const earlier = timestamp(value[earlierField]);
    const later = timestamp(value[laterField]);
    if (earlier == null || later == null || earlier <= later) continue;
    const path = `$.${earlierField}`;
    const relatedPath = `$.${laterField}`;
    issues.push({
      code: "chronology",
      path,
      relatedPath,
      message: `${path} must not be after ${relatedPath}`,
    });
  }
  return issues;
}

export class ConsumableLifecycleValidationError extends Error {
  readonly name = "ConsumableLifecycleValidationError";
  readonly collection: string;
  readonly issues: readonly ConsumableLifecycleIssue[];

  constructor(collection: string, issues: readonly ConsumableLifecycleIssue[]) {
    super(`Invalid consumable lifecycle for ${collection}: ${issues.map((issue) => issue.message).join("; ")}`);
    this.collection = collection;
    this.issues = issues;
  }
}

/** Throw when a consumable record contains chronologically impossible dates. */
export function assertConsumableLifecycle(collection: string, value: LifecycleValue): void {
  const issues = validateConsumableLifecycle(collection, value);
  if (issues.length) throw new ConsumableLifecycleValidationError(collection, issues);
}

File: apps/ios/Packages/HypoLexicon/Sources/HypoLexicon/ConsumableLifecycle.swift
import Foundation

/// A validated RFC 3339 timestamp that preserves its original wire spelling.
public struct ATProtoDate: Hashable, Sendable, Codable, CustomStringConvertible {
    public let rawValue: String
    public let date: Date

    public init(_ rawValue: String) throws {
        guard let date = Self.parse(rawValue) else {
            throw LexiconValueError.invalidDate(rawValue)
        }
        self.rawValue = rawValue
        self.date = date
    }

    public init(_ date: Date) {
        self.date = date
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        rawValue = formatter.string(from: date)
    }

    public var description: String { rawValue }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.singleValueContainer()
        try self.init(container.decode(String.self))
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }

    private static func parse(_ value: String) -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: value) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: value)
    }
}

/// Optional film-roll milestones. Their absence carries no implied date.
public struct FilmRollMilestones: Hashable, Sendable, Codable {
    public var loadedAt: ATProtoDate?
    public var partialAt: ATProtoDate?
    public var exposedAt: ATProtoDate?
    public var unloadedAt: ATProtoDate?
    public var sentToLabAt: ATProtoDate?
    public var developmentStartedAt: ATProtoDate?
    public var developedAt: ATProtoDate?
    public var receivedFromLabAt: ATProtoDate?
    public var scannedAt: ATProtoDate?
    public var archivedAt: ATProtoDate?

    public init(
        loadedAt: ATProtoDate? = nil,
        partialAt: ATProtoDate? = nil,
        exposedAt: ATProtoDate? = nil,
        unloadedAt: ATProtoDate? = nil,
        sentToLabAt: ATProtoDate? = nil,
        developmentStartedAt: ATProtoDate? = nil,
        developedAt: ATProtoDate? = nil,
        receivedFromLabAt: ATProtoDate? = nil,
        scannedAt: ATProtoDate? = nil,
        archivedAt: ATProtoDate? = nil
    ) {
        self.loadedAt = loadedAt
        self.partialAt = partialAt
        self.exposedAt = exposedAt
        self.unloadedAt = unloadedAt
        self.sentToLabAt = sentToLabAt
        self.developmentStartedAt = developmentStartedAt
        self.developedAt = developedAt
        self.receivedFromLabAt = receivedFromLabAt
        self.scannedAt = scannedAt
        self.archivedAt = archivedAt
    }
}

/// Optional chemistry milestones. Acquisition and expiry are intentionally not ordered
/// against mixing because users may acquire an opened or mixed working solution.
public struct ChemistryMilestones: Hashable, Sendable, Codable {
    public var acquiredAt: ATProtoDate?
    public var openedAt: ATProtoDate?
    public var mixedAt: ATProtoDate?
    public var replenishedAt: ATProtoDate?
    public var exhaustedAt: ATProtoDate?
    public var discardedAt: ATProtoDate?

    public init(
        acquiredAt: ATProtoDate? = nil,
        openedAt: ATProtoDate? = nil,
        mixedAt: ATProtoDate? = nil,
        replenishedAt: ATProtoDate? = nil,
        exhaustedAt: ATProtoDate? = nil,
        discardedAt: ATProtoDate? = nil
    ) {
        self.acquiredAt = acquiredAt
        self.openedAt = openedAt
        self.mixedAt = mixedAt
        self.replenishedAt = replenishedAt
        self.exhaustedAt = exhaustedAt
        self.discardedAt = discardedAt
    }
}

public struct ConsumableLifecycleIssue: Error, Hashable, Sendable {
    public let earlierField: String
    public let laterField: String

    public init(earlierField: String, laterField: String) {
        self.earlierField = earlierField
        self.laterField = laterField
    }

    public var message: String {
        "$.\(earlierField) must not be after $.\(laterField)"
    }
}

/// Cross-field chronology validation shared by the web and native authoring paths.
public enum ConsumableLifecycleValidator {
    public static func validate(_ milestones: FilmRollMilestones) -> [ConsumableLifecycleIssue] {
        validate(
            dates: [
                "loadedAt": milestones.loadedAt,
                "partialAt": milestones.partialAt,
                "exposedAt": milestones.exposedAt,
                "unloadedAt": milestones.unloadedAt,
                "sentToLabAt": milestones.sentToLabAt,
                "developmentStartedAt": milestones.developmentStartedAt,
                "developedAt": milestones.developedAt,
                "receivedFromLabAt": milestones.receivedFromLabAt,
                "scannedAt": milestones.scannedAt,
                "archivedAt": milestones.archivedAt,
            ],
            edges: [
                ("loadedAt", "partialAt"),
                ("partialAt", "exposedAt"),
                ("exposedAt", "unloadedAt"),
                ("unloadedAt", "sentToLabAt"),
                ("unloadedAt", "developmentStartedAt"),
                ("sentToLabAt", "developmentStartedAt"),
                ("developmentStartedAt", "developedAt"),
                ("developedAt", "receivedFromLabAt"),
                ("developedAt", "scannedAt"),
                ("receivedFromLabAt", "archivedAt"),
                ("scannedAt", "archivedAt"),
            ]
        )
    }

    public static func validate(_ milestones: ChemistryMilestones) -> [ConsumableLifecycleIssue] {
        validate(
            dates: [
                "acquiredAt": milestones.acquiredAt,
                "openedAt": milestones.openedAt,
                "mixedAt": milestones.mixedAt,
                "replenishedAt": milestones.replenishedAt,
                "exhaustedAt": milestones.exhaustedAt,
                "discardedAt": milestones.discardedAt,
            ],
            edges: [
                ("openedAt", "mixedAt"),
                ("mixedAt", "replenishedAt"),
                ("mixedAt", "exhaustedAt"),
                ("acquiredAt", "discardedAt"),
                ("openedAt", "discardedAt"),
                ("mixedAt", "discardedAt"),
                ("replenishedAt", "discardedAt"),
                ("exhaustedAt", "discardedAt"),
            ]
        )
    }

    private static func validate(
        dates: [String: ATProtoDate?],
        edges: [(String, String)]
    ) -> [ConsumableLifecycleIssue] {
        let pairs = transitivePairs(edges)
        return pairs.compactMap { earlier, later in
            guard let earlierDate = dates[earlier] ?? nil,
                let laterDate = dates[later] ?? nil,
                earlierDate.date > laterDate.date
            else {
                return nil
            }
            return ConsumableLifecycleIssue(earlierField: earlier, laterField: later)
        }
    }

    private static func transitivePairs(_ edges: [(String, String)]) -> [(String, String)] {
        var successors: [String: Set<String>] = [:]
        for (earlier, later) in edges {
            successors[earlier, default: []].insert(later)
        }

        var pairs: [(String, String)] = []
        for earlier in successors.keys.sorted() {
            var seen: Set<String> = []
            var pending = Array(successors[earlier] ?? []).sorted()
            while let later = pending.popLast() {
                guard seen.insert(later).inserted else { continue }
                pairs.append((earlier, later))
                pending.append(contentsOf: successors[later] ?? [])
            }
        }
        return pairs
    }
}
